fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/AgilityContext.java

```diff
@@ -299,6 +299,7 @@ private void reportFdbException(Throwable ex) {
         @Override
         public <R> CompletableFuture<R> apply(Function<FDBRecordContext, CompletableFuture<R>> function) {
             ensureOpen();
+            commitIfNeeded();
```

Comment:
This needs to be before the `ensureOpen()`. `commitIfNeeded()` results in us not having a context.
This is probably an indication that we could use a couple more tests, specifically where it calls `accept`/`apply` and then sleeps longer than the timeout, and then calls `accept`/`apply` again.
I think if you take advantage of versionstamp you could assert that they actually happen in separate transactions.
I think something like this should fail without your change and pass after your change is in:
```java
    @Test
    void name() {
        try (FDBRecordContext factoryContext = openContext()) {
            final AgilityContext agilityContext = AgilityContext.agile(factoryContext, 2, 10000);
            AtomicReference<FDBRecordContext> firstOperation = new AtomicReference<>();
            AtomicReference<FDBRecordContext> secondOperation = new AtomicReference<>();
            agilityContext.accept(context -> {
                context.ensureActive().set(key, value);
                firstOperation.set(context);
            });
            Thread.sleep(5);
            agilityContext.accept(context -> {
                context.ensureActive().set(key, value);
                secondOperation.set(context);
            });
            agilityContext.flush();
            MatcherAssert.assertThat(secondOperation.get().getCommittedVersion(), Matchers.greaterThan(firstOperation.get().getCommittedVersion()));
        }
    }
```
You could parameterize that test to run both with `accept` and `apply` like some of the other tests in `AgilityContextTest`

Reply:
IIUIC, this test may cause a "transaction too old error". As the first call will not commit and the second, even if trying to commit immediately, may already fail. Maybe we could have had: 
```
....
            Thread.sleep(4200);
            agilityContext.accept(context -> {
                context.ensureActive().set(key, value);
                secondOperation.set(context);
            });
```
But it still might be a very flaky test.

Reply:
I don't think you should have that long of a sleep. By looking at the committed versions, as I described, you are ensuring that `AgilityContext` is checking its time limit, and committing them separately.
I guess you could also just do a reference equality check to assert that the two operations get different contexts, that might be better, as it depends on less fdb semantics.

Reply:
✅